File: android/app/src/main/java/com/bobril/bobriln/Decoder.java
package com.bobril.bobriln;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

enum EncodedType {
    Number0(0),
    Number63(63),
    String0(64),
    String63(127),
    Map0(128),
    Map31(159),
    Array0(160),
    Array31(191),
    PositiveInt1(192),
    PositiveInt4(195),
    NegativeInt1(196),
    NegativeInt4(199),
    String1(200),
    String4(203),
    Map1(204),
    Map4(207),
    Array1(208),
    Array4(211),
    Double(212),
    False(213),
    True(214),
    Null(215),
    EndOfBlock(216),
    NumberM32(224),
    NumberM1(255);

    public final int val;

    EncodedType(int param) {
        this.val = param;
    }
}

enum PeekType {
    Int,
    Double,
    String,
    Boolean,
    Null,
    ArrayStart,
    ArrayEnd,
    MapStart,
    MapEnd,
    EndOfBlock,
    EOF
}

enum StateType {
    Array,
    MapKey,
    MapValue
}

public class Decoder {
    byte[] _buf;
    ByteBuffer _bbuf;
    int[] _stack;
    int _stackPos;
    PeekType _peekType;
    String _peekString;
    StringBuilder _sb;
    int _peekInt;
    double _peekDouble;
    boolean _peekBoolean;

    public Decoder() {
        this._stack = new int[4];
        this._stackPos = 0;
        this._sb = new StringBuilder();
    }

    public void initFromString(String value) {
        int len = value.length();
        if (_buf == null || _buf.length < len) _buf = new byte[len];
        for (int i = 0; i < len; i++) {
            _buf[i] = (byte) value.charAt(i);
        }
        _bbuf = ByteBuffer.wrap(_buf, 0, len);
        _bbuf.order(ByteOrder.LITTLE_ENDIAN);
        this._stackPos = 0;
        this._peekType = PeekType.EOF;
        this._peekString = null;
        this.next();
    }

    private void setPeek(PeekType type, String lit, int ofs) {
        this._peekType = type;
        this._peekString = lit;
        this._bbuf.position(ofs);
    }

    private void setPeek(PeekType type, int lit, int ofs) {
        this._peekType = type;
        this._peekInt = lit;
        this._peekString = null;
        this._bbuf.position(ofs);
    }

    private void setPeek(PeekType type, double lit, int ofs) {
        this._peekType = type;
        this._peekDouble = lit;
        this._peekString = null;
        this._bbuf.position(ofs);
    }

    private void setPeek(PeekType type, boolean lit, int ofs) {
        this._peekType = type;
        this._peekBoolean = lit;
        this._peekString = null;
        this._bbuf.position(ofs);
    }

    private void throwEOF() throws Error {
        throw new Error("EOF too early");
    }

    private void parseString(int strLen, int ofs) {
        StringBuilder sb = this._sb;
        sb.setLength(0);
        ByteBuffer buf = this._bbuf;
        int len = buf.limit();
        while (strLen-- > 0) {
            if (ofs >= len) this.throwEOF();
            int b = 255 & buf.get(ofs++);
            if (b < 0x80) {
                sb.append((char) b);
            } else if (b < 0xc0) {
                if (ofs >= len) this.throwEOF();
                sb.append((char) (((b - 0x80) << 8) + (255 & buf.get(ofs++))));
            } else {
                if (ofs + 1 >= len) this.throwEOF();
                sb.append((char) (((255 & buf.get(ofs)) << 8) + (255 & buf.get(ofs + 1))));
                ofs += 2;
            }
        }
        this.setPeek(PeekType.String, sb.toString(), ofs);
    }

    private int readVarLenInt(int lenBytesM1, int ofs) {
        ByteBuffer buf = this._bbuf;
        if (lenBytesM1 == 0) {
            buf.position(ofs + 1);
            return 255 & buf.get(ofs);
        }
        if (lenBytesM1 == 1) {
            buf.position(ofs + 2);
            return 256 + ((255 & buf.get(ofs)) << 8) + (255 & buf.get(ofs + 1));
        }
        if (lenBytesM1 == 2) {
            buf.position(ofs + 3);
            return 0x10100 + ((255 & buf.get(ofs)) << 16) + ((255 & buf.get(ofs + 1)) << 8) + (255 & buf.get(ofs + 2));
        }
        buf.position(ofs + 4);
        return 0x1010100 + ((255 & buf.get(ofs)) << 24) + ((255 & buf.get(ofs + 1)) << 16) + ((255 & buf.get(ofs + 2)) << 8) + (255 & buf.get(ofs + 3));
    }

    public void next() {
        switch (this._peekType) {
            case ArrayStart: {
                if (this._stack[this._stackPos - 1]-- == 0) {
                    this._peekType = PeekType.ArrayEnd;
                    this._stackPos -= 2;
                    return;
                }
                break;
            }
            case MapStart: {
                if (this._stack[this._stackPos - 1]-- == 0) {
                    this._peekType = PeekType.MapEnd;
                    this._stackPos -= 2;
                    return;
                }
                break;
            }
            default: {
                int[] s = this._stack;
                int sl = this._stackPos;
                if (sl > 0) {
                    if (s[sl - 2] == StateType.Array.ordinal()) {
                        if (s[sl - 1]-- == 0) {
                            this._peekType = PeekType.ArrayEnd;
                            this._peekString = null;
                            this._stackPos -= 2;
                            return;
                        }
                    } else if (s[sl - 2] == StateType.MapKey.ordinal()) {
                        s[sl - 2] = StateType.MapValue.ordinal();
                    } else {
                        s[sl - 2] = StateType.MapKey.ordinal();
                        if (s[sl - 1]-- == 0) {
                            this._peekType = PeekType.MapEnd;
                            this._peekString = null;
                            this._stackPos -= 2;
                            return;
                        }
                    }
                }
                break;
            }
        }
        ByteBuffer buf = this._bbuf;
        int ofs = buf.position();
        int len = buf.limit();
        if (ofs >= len) {
            this.setPeek(PeekType.EOF, null, ofs);
            return;
        }
        int b = 255 & buf.get(ofs++);
        if (b <= EncodedType.Number63.val) {
            this.setPeek(PeekType.Int, b, ofs);
            return;
        }
        if (b >= EncodedType.NumberM32.val) {
            this.setPeek(PeekType.Int, b - 256, ofs);
            return;
        }
        if (b <= EncodedType.String63.val) {
            int l = b - EncodedType.String0.val;
            if (l == 0) {
                this.setPeek(PeekType.String, "", ofs);
                return;
            }
            this.parseString(l, ofs);
            return;
        }
        if (b <= EncodedType.Map31.val) {
            this.stackPush(StateType.MapKey, b - EncodedType.Map0.val);
            this.setPeek(PeekType.MapStart, null, ofs);
            return;
        }
        if (b <= EncodedType.Array31.val) {
            this.stackPush(StateType.Array, b - EncodedType.Array0.val);
            this.setPeek(PeekType.ArrayStart, null, ofs);
            return;
        }
        switch (b) {
            case 192 /* EncodedType.PositiveInt1 */:
            case 193 /* EncodedType.PositiveInt1 + 1 */:
            case 194 /* EncodedType.PositiveInt1 + 2 */:
            case 195 /* EncodedType.PositiveInt4 */: {
                this._peekType = PeekType.Int;
                this._peekInt = 64 + this.readVarLenInt(b - EncodedType.PositiveInt1.val, ofs);
                break;
            }
            case 196 /* EncodedType.NegativeInt1 */:
            case 197 /* EncodedType.NegativeInt1 + 1 */:
            case 198 /* EncodedType.NegativeInt1 + 2 */:
            case 199 /* EncodedType.NegativeInt4 */: {
                this._peekType = PeekType.Int;
                this._peekInt = -33 - this.readVarLenInt(b - EncodedType.NegativeInt1.val, ofs);
                break;
            }
            case 200 /* EncodedType.String1 */:
            case 201 /* EncodedType.String1 + 1 */:
            case 202 /* EncodedType.String1 + 2 */:
            case 203 /* EncodedType.String4 */: {
                int l = 64 + this.readVarLenInt(b - EncodedType.String1.val, ofs);
                this.parseString(l, this._bbuf.position());
                break;
            }
            case 204 /* EncodedType.Map1 */:
            case 205 /* EncodedType.Map1 + 1 */:
            case 206 /* EncodedType.Map1 + 2 */:
            case 207 /* EncodedType.Map4 */: {
                int l = 32 + this.readVarLenInt(b - EncodedType.Map1.val, ofs);
                this.stackPush(StateType.MapKey, l);
                this._peekType = PeekType.MapStart;
                break;
            }
            case 208 /* EncodedType.Array1 */:
            case 209 /* EncodedType.Array1 + 1 */:
            case 210 /* EncodedType.Array1 + 2 */:
            case 211 /* EncodedType.Array4 */: {
                int l = 32 + this.readVarLenInt(b - EncodedType.Array1.val, ofs);
                this.stackPush(StateType.Array, l);
                this._peekType = PeekType.ArrayStart;
                break;
            }
            case 212 /* EncodedType.Double */: {
                this.setPeek(PeekType.Double, buf.getDouble(ofs), ofs + 8);
                break;
            }
            case 213 /* EncodedType.False */: {
                this.setPeek(PeekType.Boolean, false, ofs);
                break;
            }
            case 214 /* EncodedType.True */: {
                this.setPeek(PeekType.Boolean, true, ofs);
                break;
            }
            case 215 /* EncodedType.Null */: {
                this.setPeek(PeekType.Null, null, ofs);
                break;
            }
            case 216 /* EncodedType.EndOfBlock */: {
                this.setPeek(PeekType.EndOfBlock, null, ofs);
                break;
            }
            default:
                throw new Error("Corrupted");
        }
    }

    private void stackPush(StateType type, int idx) {
        if (this._stackPos + 2 > this._stack.length) {
            this._stack = Arrays.copyOf(this._stack, this._stack.length * 2);
        }
        this._stack[this._stackPos++] = type.ordinal();
        this._stack[this._stackPos++] = idx;
    }

    public boolean isAnyEnd() {
        return (this._peekType == PeekType.EOF) || (this._peekType == PeekType.EndOfBlock);
    }

    public boolean isEOF() {
        return this._peekType == PeekType.EOF;
    }

    public int readInt() {
        if (_peekType != PeekType.Int)
            throw new Error("Cannot read int from " + _peekType.name());
        int res = _peekInt;
        next();
        return res;
    }

    public double readDouble() {
        if (_peekType != PeekType.Int && _peekType != PeekType.Double)
            throw new Error("Cannot read double from " + _peekType.name());
        double res = _peekType == PeekType.Double ? _peekDouble : _peekInt;
        next();
        return res;
    }

    public boolean readBoolean() {
        if (_peekType != PeekType.Boolean)
            throw new Error("Cannot read boolean from " + _peekType.name());
        boolean res = _peekBoolean;
        next();
        return res;
    }

    public String readString() {
        if (_peekType != PeekType.String)
            throw new Error("Cannot read string from " + _peekType.name());
        String res = _peekString;
        next();
        return res;
    }

    public Object readAny() {
        switch (_peekType) {
            case EOF:
            case EndOfBlock: {
                throwEOF();
            }
            case Boolean: {
                Object res = _peekBoolean ? Boolean.TRUE : Boolean.FALSE;
                next();
                return res;
            }
            case Null: {
                next();
                return null;
            }
            case Int: {
                Object res = _peekInt;
                next();
                return res;
            }
            case Double: {
                Object res = _peekDouble;
                next();
                return res;
            }
            case String: {
                Object res = _peekString;
                next();
                return res;
            }
            case ArrayEnd:
            case MapEnd: {
                throw new Error("Reading something end");
            }
            case ArrayStart: {
                List<Object> res = new ArrayList<Object>();
                next();
                while (_peekType != PeekType.ArrayEnd) {
                    res.add(readAny());
                }
                next();
                return res;
            }
            case MapStart: {
                HashMap<String, Object> res = new HashMap<>();
                next();
                while (_peekType != PeekType.MapEnd) {
                    if (_peekType != PeekType.String)
                        throw new Error("Object key is not string");
                    String key = _peekString;
                    _peekString = null;
                    next();
                    res.put(key, readAny());
                }
                next();
                return res;
            }
            default:
                throw new Error("Unexpected peekType " + _peekType.name());
        }
    }
}

File: decoder.ts
import { EncodedType } from './decEncCommon';

export const enum PeekType {
    Number = 0,
    String = 1,
    Boolean = 2,
    Null = 3,
    ArrayStart = 4,
    ArrayEnd = 5,
    MapStart = 6,
    MapEnd = 7,
    EndOfBlock = 8,
    EOF = 9
}

const enum StateType {
    Array,
    MapKey,
    MapValue
}

export class Decoder {
    buf: Uint8Array;
    dv: DataView;
    ofs: number;
    len: number;
    stack: number[];
    peekType: PeekType;
    peekLiteral: number | string | boolean;

    constructor() {
        this.buf = null;
        this.ofs = 0;
        this.len = 0;
        this.stack = [];
        this.dv = null;
        this.peekType = PeekType.EOF;
        this.peekLiteral = undefined;
    }

    initFromLatin1String(value: string) {
        const len = value.length;
        if (this.len < len) {
            this.buf = new Uint8Array(len);
        }
        const buf = this.buf;
        for (let i = 0; i < len; i++) {
            buf[i] = value.charCodeAt(i);
        }
        this.ofs = 0;
        this.len = len;
        this.dv = null;
        this.stack.length = 0;
        this.peekType = PeekType.EOF;
        this.peekLiteral = undefined;
        this.next();
    }

    private setPeek(type: PeekType, lit: number | string | boolean, ofs: number) {
        this.peekType = type;
        this.peekLiteral = lit;
        this.ofs = ofs;
    }

    private throwEOF() {
        throw new Error("EOF too early");
    }

    private getDv(): DataView {
        if (this.dv !== null) return this.dv;
        this.dv = new DataView(this.buf.buffer);
        return this.dv;
    }

    private parseString(strLen: number, ofs: number) {
        let batchLen = Math.min(4096, strLen);
        let sb = new Uint16Array(batchLen);
        let res = "";
        let buf = this.buf;
        let len = this.len;
        while (strLen > 0) {
            if (strLen < batchLen) batchLen = strLen;
            let oofs = 0;
            while (oofs < batchLen) {
                if (ofs >= len) this.throwEOF();
                let b = buf[ofs++];
                if (b < 0x80) {
                    sb[oofs++] = b;
                } else if (b < 0xc0) {
                    if (ofs >= len) this.throwEOF();
                    sb[oofs++] = ((b - 0x80) << 8) + buf[ofs++];
                } else {
                    if (ofs + 1 >= len) this.throwEOF();
                    sb[oofs++] = (buf[ofs] << 8) + buf[ofs + 1];
                    ofs += 2;
                }
            }
            res += String.fromCharCode.apply(null, sb.subarray(0, batchLen));
            strLen -= batchLen;
        }
        this.setPeek(PeekType.String, res, ofs);
    }

    private readVarLenInt(lenBytesM1: number, ofs: number): number {
        if (lenBytesM1 === 0) {
            this.ofs = ofs + 1;
            return this.buf[ofs];
        }
        if (lenBytesM1 === 1) {
            this.ofs = ofs + 2;
            let buf = this.buf;
            return 256 + (buf[ofs] << 8) + buf[ofs + 1];
        }
        if (lenBytesM1 === 2) {
            this.ofs = ofs + 3;
            let buf = this.buf;
            return 0x10100 + (buf[ofs] << 16) + (buf[ofs + 1] << 8) + buf[ofs + 2];
        }
        this.ofs = ofs + 4;
        let buf = this.buf;
        return 0x1010100 + (buf[ofs] << 24) + (buf[ofs + 1] << 16) + (buf[ofs + 2] << 8) + buf[ofs + 3];
    }

    next() {
        switch (this.peekType) {
            case PeekType.ArrayStart: {
                if (this.stack[this.stack.length - 1]-- === 0) {
                    this.peekType = PeekType.ArrayEnd;
                    this.stack.length -= 2;
                    return;
                }
                break;
            }
            case PeekType.MapStart: {
                if (this.stack[this.stack.length - 1]-- === 0) {
                    this.peekType = PeekType.MapEnd;
                    this.stack.length -= 2;
                    return;
                }
                break;
            }
            default: {
                let s = this.stack;
                let sl = s.length;
                if (sl > 0) {
                    if (s[sl - 2] === StateType.Array) {
                        if (s[sl - 1]-- === 0) {
                            this.peekType = PeekType.ArrayEnd;
                            this.peekLiteral = undefined;
                            s.length -= 2;
                            return;
                        }
                    } else if (s[sl - 2] === StateType.MapKey) {
                        s[sl - 2] = StateType.MapValue;
                    } else {
                        s[sl - 2] = StateType.MapKey;
                        if (s[sl - 1]-- === 0) {
                            this.peekType = PeekType.MapEnd;
                            this.peekLiteral = undefined;
                            s.length -= 2;
                            return;
                        }
                    }
                }
                break;
            }
        }
        let buf = this.buf;
        let ofs = this.ofs;
        let len = this.len;
        if (ofs >= len) {
            this.setPeek(PeekType.EOF, undefined, ofs);
            return;
        }
        let b = buf[ofs++];
        if (b <= EncodedType.Number63) {
            this.setPeek(PeekType.Number, b, ofs);
            return;
        }
        if (b >= EncodedType.NumberM32) {
            this.setPeek(PeekType.Number, b - 256, ofs);
            return;
        }
        if (b <= EncodedType.String63) {
            let l = b - EncodedType.String0;
            if (l === 0) {
                this.setPeek(PeekType.String, "", ofs);
                return;
            }
            this.parseString(l, ofs);
            return;
        }
        if (b <= EncodedType.Map31) {
            this.stack.push(StateType.MapKey, b - EncodedType.Map0);
            this.setPeek(PeekType.MapStart, undefined, ofs);
            return;
        }
        if (b <= EncodedType.Array31) {
            this.stack.push(StateType.Array, b - EncodedType.Array0);
            this.setPeek(PeekType.ArrayStart, undefined, ofs);
            return;
        }
        switch (b) {
            case EncodedType.PositiveInt1:
            case EncodedType.PositiveInt1 + 1:
            case EncodedType.PositiveInt1 + 2:
            case EncodedType.PositiveInt4: {
                this.peekType = PeekType.Number;
                this.peekLiteral = 64 + this.readVarLenInt(b - EncodedType.PositiveInt1, ofs);
                break;
            }
            case EncodedType.NegativeInt1:
            case EncodedType.NegativeInt1 + 1:
            case EncodedType.NegativeInt1 + 2:
            case EncodedType.NegativeInt4: {
                this.peekType = PeekType.Number;
                this.peekLiteral = -33 - this.readVarLenInt(b - EncodedType.NegativeInt1, ofs);
                break;
            }
            case EncodedType.String1:
            case EncodedType.String1 + 1:
            case EncodedType.String1 + 2:
            case EncodedType.String4: {
                let l = 64 + this.readVarLenInt(b - EncodedType.String1, ofs);
                this.parseString(l, this.ofs);
                break;
            }
            case EncodedType.Map1:
            case EncodedType.Map1 + 1:
            case EncodedType.Map1 + 2:
            case EncodedType.Map4: {
                let l = 32 + this.readVarLenInt(b - EncodedType.Map1, ofs);
                this.stack.push(StateType.MapKey, l);
                this.peekType = PeekType.MapStart;
                this.peekLiteral = undefined;
                break;
            }
            case EncodedType.Array1:
            case EncodedType.Array1 + 1:
            case EncodedType.Array1 + 2:
            case EncodedType.Array4: {
                let l = 32 + this.readVarLenInt(b - EncodedType.Array1, ofs);
                this.stack.push(StateType.Array, l);
                this.peekType = PeekType.ArrayStart;
                this.peekLiteral = undefined;
                break;
            }
            case EncodedType.Double: {
                this.setPeek(PeekType.Number, this.getDv().getFloat64(ofs, true), ofs + 8);
                break;
            }
            case EncodedType.False: {
                this.setPeek(PeekType.Boolean, false, ofs);
                break;
            }
            case EncodedType.True: {
                this.setPeek(PeekType.Boolean, true, ofs);
                break;
            }
            case EncodedType.Null: {
                this.setPeek(PeekType.Null, null, ofs);
                break;
            }
            case EncodedType.EndOfBlock: {
                this.setPeek(PeekType.EndOfBlock, null, ofs);
                break;
            }
            default:
                throw new Error("Corrupted");
        }
    }

    isAnyEnd() {
        return (this.peekType === PeekType.EOF) || (this.peekType === PeekType.EndOfBlock);
    }

    isEOF() {
        return this.peekType === PeekType.EOF;
    }

    readAny(): any {
        switch (this.peekType) {
            case PeekType.EOF:
            case PeekType.EndOfBlock: {
                this.throwEOF();
            }
            case PeekType.Boolean:
            case PeekType.Null:
            case PeekType.Number:
            case PeekType.String: {
                let res = this.peekLiteral;
                this.next();
                return res;
            }
            case PeekType.ArrayEnd:
            case PeekType.MapEnd: {
                throw new Error("Reading something end");
            }
            case PeekType.ArrayStart: {
                let res: any[] = [];
                this.next();
                while (this.peekType !== PeekType.ArrayEnd) {
                    res.push(this.readAny());
                }
                this.next();
                return res;
            }
            case PeekType.MapStart: {
                let res = Object.create(null);
                this.next();
                while (this.peekType !== PeekType.MapEnd) {
                    if (this.peekType !== PeekType.String)
                        throw new Error("Object key is not string");
                    let key = <string>this.peekLiteral;
                    this.next();
                    res[key] = this.readAny();
                }
                this.next();
                return res;
            }
            default: throw new Error("Unexpected peekType " + this.peekType);
        }
    }
}
